src/printer/code_printer.cpp

```diff
@@ -49,41 +49,59 @@ void CodePrinter::restart_block(std::string const& expression) {
 }
 
 void CodePrinter::add_indent() {
-    *result << std::string(indent_level * NUM_SPACES, ' ');
-}
-
-void CodePrinter::add_text(const std::string& text) {
-    *result << text;
-}
-
-void CodePrinter::add_line(const std::string& text, int num_new_lines) {
-    add_indent();
-    *result << text;
-    add_newline(num_new_lines);
+    for (std::size_t i = 0; i < indent_level * NUM_SPACES; ++i) {
+        *result << ' ';
+    }
 }
 
 void CodePrinter::add_multi_line(const std::string& text) {
-    auto lines = stringutils::split_string(text, '\n');
+    const auto& lines = stringutils::split_string(text, '\n');
+
+    int prefix_length{};
+    int start_line{};
+    while (start_line < lines.size()) {
+        const auto& line = lines[start_line];
+        // skip first empty line, if any
+        if (line.empty()) {
+            ++start_line;
+            continue;
+        }
+        // The common indentation of all blocks if the number of spaces
+        // at the beginning of the first non-empty line.
+        for (auto line_it = line.begin(); line_it != line.end() && *line_it == ' '; ++line_it) {
+            prefix_length += 1;
+        }
+        break;
+    }
+
     for (const auto& line: lines) {
-        add_line(line);
+        if (line.size() < prefix_length) {
+            // ignore lines made of ' ' characters
+            if (std::find_if_not(line.begin(), line.end(), [](char c) { return c == ' '; }) !=
+                line.end()) {
+                throw std::invalid_argument("Indentation mismatch");
+            }
+        } else {
+            add_line(line.substr(prefix_length));
+        }
     }
 }
 
-void CodePrinter::add_newline(int n) {
-    for (int i = 0; i < n; i++) {
-        *result << std::endl;
+void CodePrinter::add_newline(std::size_t n) {
+    for (std::size_t i{}; i < n; ++i) {
+        *result << '\n';
     }
 }
 
-void CodePrinter::end_block(int num_newlines) {
+void CodePrinter::pop_block(int num_newlines) {
```

Comment:
I think that `1` for `num_newlines` should be by default.

If you want `0` use `chain_block`

If you want more use `add_newline` yourself.

Reply:
What do you suggest? To remove the `num_newlines` parameter right?
Good be a nice improvement indeed. Let's try this in a separate PR if you are ok with this!

Reply:
Yes, this is my idea. Ok for next PR of course.

Reply:
Actually implemented in latest change
d9bd59e9c56555